File: core/src/main/java/forpleuvoir/dhwu1a/core/event/bot/BotEvent.java
package forpleuvoir.dhwu1a.core.event.bot;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.gson.JsonObject;
import com.google.gson.annotations.SerializedName;
import forpleuvoir.dhwu1a.core.event.base.Dhwu1aEvent;
import forpleuvoir.dhwu1a.core.util.JsonUtil;
import forpleuvoir.dhwu1a.core.util.ReflectionUtil;

import java.util.function.Consumer;

import static forpleuvoir.dhwu1a.core.common.ApiKey.TYPE;

/**
 * Bot事件
 *
 * @author forpleuvoir
 * <p>#project_name dhwu1a
 * <p>#package forpleuvoir.dhwu1a.core.event.bot
 * <p>#class_name BotEvent
 * <p>#create_time 2021/7/3 20:25
 */
public abstract class BotEvent extends Dhwu1aEvent {
    /**
     * 事件类型
     */
    @SerializedName(TYPE)
    public final BotEventType type;

    protected BotEvent(BotEventType type) {
        this.type = type;
    }

    public static BotEvent parse(JsonObject object) {
        var type = BotEventType.valueOf(object.get(TYPE).getAsString());
        return (BotEvent) JsonUtil.gson.fromJson(object, type.getClazz()).initialize();
    }

    @Override
    @SuppressWarnings({"rawtypes", "unchecked"})
    public void broadcastHandle(
            ImmutableMap<Class<? extends Dhwu1aEvent>, ImmutableList<Consumer<? extends Dhwu1aEvent>>> eventListeners
    ) {
        super.broadcastHandle(eventListeners);
        if (ReflectionUtil.isExtended(this.getClass(), BotEvent.class)) {
            if (eventListeners.containsKey(BotEvent.class)) {
                for (Consumer listener : eventListeners.get(BotEvent.class)) {
                    listener.accept(this);
                }
            }
        }
    }

    @Override
    public String toPlainText() {
        return String.format("%s[%d]", type, bot.id);
    }
}

File: core/src/main/kotlin/forpleuvoir/dhwu1a/core/event/bot/BotEvent.kt
package forpleuvoir.dhwu1a.core.event.bot

import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import forpleuvoir.dhwu1a.core.common.TYPE
import forpleuvoir.dhwu1a.core.event.base.Dhwu1aEvent
import forpleuvoir.dhwu1a.core.util.JsonUtil
import forpleuvoir.dhwu1a.core.util.ReflectionUtil
import java.util.function.Consumer

/**
 * Bot事件
 *
 * @author forpleuvoir
 *
 * #project_name dhwu1a
 *
 * #package forpleuvoir.dhwu1a.core.event.bot
 *
 * #class_name BotEvent
 *
 * #create_time 2021/7/3 20:25
 */
abstract class BotEvent protected constructor(
    /**
     * 事件类型
     */
    @field:SerializedName(TYPE) val type: BotEventType
) : Dhwu1aEvent() {

    override fun broadcastHandle(
        eventListeners: ImmutableMap<Class<out Dhwu1aEvent>, ImmutableList<Consumer<in Dhwu1aEvent>>>
    ) {
        super.broadcastHandle(eventListeners)
        if (ReflectionUtil.isExtended(this.javaClass, BotEvent::class.java)) {
            if (eventListeners.containsKey(BotEvent::class.java)) {
                for (listener in eventListeners[BotEvent::class.java]!!) {
                    listener.accept(this)
                }
            }
        }
    }

    override fun toPlainText(): String {
        return String.format("%s[%d]", type, bot!!.id)
    }

    companion object {
        @JvmStatic
        fun parse(`object`: JsonObject): BotEvent {
            val type = BotEventType.valueOf(`object`[TYPE].asString)
            return JsonUtil.gson.fromJson(`object`, type.clazz)!!.initialize() as BotEvent
        }
    }
}